### DISCUSSION TITLE
Why useEffect in useCallbackRef?

### DISCUSSION THREAD

**Discussion by pshdev1030** [radix-ui/primitives#2114] [Ideas]
## Question

In [useCallbackRef](https://github.com/radix-ui/primitives/blob/main/packages/react/use-callback-ref/src/useCallbackRef.tsx), I think it would be safer to use useLayoutEffect because useLayoutEffect runs before print, so is there a reason to use it?

I think there will be little impact on rendering performance, is that the reason?

```ts
import * as React from 'react';

function useCallbackRef<T extends (...args: any[]) => any>(callback: T | undefined): T {
  const callbackRef = React.useRef(callback);

  React.useLayoutEffect(() => {
    callbackRef.current = callback;
  });

  return React.useMemo(() => ((...args) => callbackRef.current?.(...args)) as T, []);
}

export { useCallbackRef };
```

**Comment by benoitgrelard** [comment]
There shouldn't be any reason to use `useLayoutEffect` over `useEffect` here. `useLayoutEffect` fires before the browser repaints the screen, so it's only really needed for things that depends on layout, etc like measuring elements to position things correctly. Otherwise you might see a flicker of an old position in a previous paint.
Here we are merely keeping a callback in sync, so that's irrelevant.

**Reply by pshdev1030** [reply]
I misunderstood. Thank you!
